zulip_bots/zulip_bots/test_lib.py

```diff
@@ -26,130 +26,135 @@
-        self.message_handler.handle_message(message, self.mock_bot_handler)
-
-        # Check if the bot is sending a message via `send_message` function.
-        # Where response is a dictionary here.
-        instance = self.MockClass.return_value
-        if expected_method == "send_message":
-            instance.send_message.assert_called_with(response)
-        else:
-            instance.send_reply.assert_called_with(message, response['content'])
-
-    @contextmanager
-    def mock_config_info(self, config_info):
-        # type: (Dict[str, str]) -> Any
-        self.MockClass.return_value.get_config_info.return_value = config_info
-        yield
-        self.MockClass.return_value.get_config_info.return_value = None
-
-    @contextmanager
-    def mock_http_conversation(self, test_name):
-        # type: (str) -> Any
-        """
-        Use this context manager to mock and verify a bot's HTTP
-        requests to the third-party API (and provide the correct
-        third-party API response. This allows us to test things
-        that would require the Internet without it).
-        """
-        assert test_name is not None
-        base_path = os.path.realpath(os.path.join(os.path.dirname(
-            os.path.abspath(__file__)), 'bots', self.bot_name, 'fixtures'))
-        http_data_path = os.path.join(base_path, '{}.json'.format(test_name))
-        with open(http_data_path, 'r') as http_data_file:
-            http_data = json.load(http_data_file)
-            http_request = http_data.get('request')
-            http_response = http_data.get('response')
-            http_headers = http_data.get('response-headers')
-            with patch('requests.get') as mock_get:
-                mock_result = mock.MagicMock()
-                mock_result.json.return_value = http_response
-                if 'text' in http_response:
-                    mock_result.text = http_response.get('text', None)
-                mock_result.status_code = http_headers.get('status', 200)
-                mock_result.ok.return_value = http_headers.get('ok', True)
-                mock_get.return_value = mock_result
-                yield
-                if 'params' in http_request:
-                    params = http_request.get('params', None)
-                    mock_get.assert_called_with(http_request['api_url'], params=params)
-                elif 'headers' in http_request:
-                    headers = http_request.get('headers', None)
-                    mock_get.assert_called_with(http_request['api_url'], headers=headers)
-                else:
-                    mock_get.assert_called_with(http_request['api_url'])
-
-    def assert_bot_response(self, message, response, expected_method):
-        # type: (Dict[str, Any], Dict[str, Any], str) -> None
-        # Strictly speaking, this function is not needed anymore,
-        # kept for now for legacy reasons.
-        self.call_request(message, expected_method, response)
+class BotTestCase:
```

Comment:
I think it might be better to keep the existing code structure as `BotTestCaseBase`, and then have this double-layer class just inherit from that and contain a bit of code for the actual tests. That avoids indenting everything twice, and also provides an interface for a bot to inherit from BotTestCaseBase if it doesn't want the common usage tests.

Does that make sense?

Reply:
Yeah, that'd make it a lot nicer!